nondex-instrumentation/src/main/java/edu/illinois/nondex/instr/Instrumenter.java

```diff
@@ -87,17 +139,57 @@ private static void instrumentClass(String className,
         outZip.closeEntry();
     }
 
-    public static final void instrument(String rtJar, String outJar) throws IOException {
+    private static void filterCachedFrom(HashSet<String> toShuffle, HashSet<String> toCopy,
+                                         ZipFile rt, ZipFile outJarZipFile)
+            throws IOException, NoSuchAlgorithmException {
+        Iterator<String> it = toShuffle.iterator();
+        while (it.hasNext()) {
+            String cl = it.next();
+            InputStream clInputStream = rt.getInputStream(rt.getEntry(cl));
+            InputStream md5InputStream = outJarZipFile.getInputStream(outJarZipFile.getEntry(cl + ".md5"));
+            if (Arrays.equals(Instrumenter.toMd5(clInputStream), Instrumenter.readAllBytes(md5InputStream))) {
+                it.remove();
+                toCopy.add(cl);
+            }
+        }
+    }
+
+    private static void filterCached(String outJar, ZipFile rt) throws IOException, NoSuchAlgorithmException {
+        if (new File(outJar).exists()) {
+            ZipFile outJarZipFile = new ZipFile(outJar);
+            Instrumenter.filterCachedFrom(Instrumenter.classesToShuffle, Instrumenter.classesToCopy, rt, outJarZipFile);
+            Instrumenter.filterCachedFrom(Instrumenter.specialClassesToShuffle, Instrumenter.specialClassesToCopy,
+                    rt, outJarZipFile);
+        }
+    }
+
+    private static void copyToJar(String cl, ZipFile oldJar, ZipOutputStream outZip) throws IOException {
+        InputStream clInputStream = oldJar.getInputStream(oldJar.getEntry(cl));
+        byte[] arr = Instrumenter.readAllBytes(clInputStream);
+
+        ZipEntry entry = new ZipEntry(cl);
+        outZip.putNextEntry(entry);
+        outZip.write(arr, 0, arr.length);
+        outZip.closeEntry();
+    }
+
+    public static final void instrument(String rtJar, String outJar)
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method instrument() has an NPath complexity of 288](https://www.codacy.com/app/gyori/NonDex/file/2716321334/issues/source?bid=3322087&fileBranchId=3343849#l176)